rfcs/0029-smart-contracts.md

```diff
@@ -0,0 +1,420 @@
+    is placed in consensus.
+* Need to implement or integrate a WASM interpreter.
+  - Gas metering may require wide-reaching changes if we adapt an existing
+    interpreter.
+* Consensus nodes need to do an amount of unpaid work to verify the gas usage.
+* Constraint system programs must be written in a language that compiles down
+  to WASM.
+  - This usually means that the language relies on manual memory management,
+    which will make it hard to use for some developers.
+* Will need to write WASM backends for our current snark langauge tooling, if
+  we want it to be usable for smart contracts.
+* The size of the constraint system is restricted by the transaction snark.
+* Verification of universal snarks is (generally) more expensive.
+* The smart contract author needs to communicate the program to generate
+  witnesses off-chain.
+
+#### Benefits
+[constraints-only-wasm-benefits]: #benefits-1
+
+* Far more insight into the predicate's behaviour.
+* Less snark-specific knowledge is necessary for developers.
+* The smart contract author doesn't need to generate keys.
+  - This reduces computational costs and data transfer costs (for the large
+    proving keys).
+* Smart contracts may be copied directly between accounts.
+* There is a good amount of WASM tooling under active development.
+* Existing WASM interpreters generally provide good sandboxing.
+
+### Witness generation + proving on chain: store WASM blobs for generating constraints and witnesses in accounts
+
+#### Drawbacks
+
+* All of the drawbacks of constraints-only WASM (see
+  [above][constraints-only-wasm-drawbacks]). Exceptions:
+  - The sender of the transaction no longer needs to generate a proof.
+  - The smart contract author does not need to participate in off-chain
+    communications.
+* The proof model is more complicated.
+  - Additional public input need to be available, to give the sender control
+    over the proof.
+* Consensus nodes need to do more unpaid work to verify the gas usage.
+* More failure cases to handle in the gas model.
+
+#### Benefits
+
+* All of the benefits for constraints-only WASM (see
+  [above][constraints-only-wasm-benefits]).
+* Transaction sender does not need to perform an expensive computation
+  (proving).
+  - This allows low-power devices to trigger smart contracts.
+* All information needed for evaluating the contract is on-chain.
+* User experience is better.
+  - Smart contract authors only send the minimum necessary information: the
+    constraint system generator and the witness generator.
+  - Transation senders only send the transaction and any other necessary
+    information. No extra computation or data is needed.
+
+### Witness generation + proving on chain: store snarky programs in accounts
+
+#### Drawbacks
+
+* Ill-defined: it isn't clear how snarky programs should be encoded when stored
+  on the chain.
+* Evaluation model is far more complex. In particular:
```

Comment:
This makes sense... if we fix the platform the code is being compiled + run on, maybe there is some way to get around this? I know we can use ulimit for memory usage, perhaps there is some way to limit the number of instructions a binary can run on linux before it dies (the Hardware Performance Counter API)? Kind of inflexible, but trying to think through what we can do to hack together a solution we could get done in time.